tests/virt_autotest/sriov_pci_passthrough.pm

```diff
@@ -0,0 +1,381 @@
+sub get_sriov_ethernet_devices () {
+
+  #get the BDF of the ethernet devices with SR-IOV
+  my $nic_devices = script_output "lspci | grep Ethernet | grep -v 'Virtual Function' | cut -d ' ' -f1" ;
+  my @nic_devices = split ("\n", $nic_devices);
+  my @sriov_devices;
+  foreach (@nic_devices){
+    if (! script_run "lspci -v -s $_ | grep -q 'SR-IOV'") {
+      push @sriov_devices, $_;
+    }
+  }
+  record_soft_failure "Error: there is no ethernet devices with SR-IOV in the host and test aborts!" if ! $#sriov_devices;
+  
+  return @sriov_devices;
+}
+
+#enable 8 virtual functions for the specified physical functions of the SR-IOV network device 
+sub enable_vf(){
+  my @pfs = @_;
+  
+  # get the network device drivers
+  my (@drivers, $pf, $driver);
+  foreach $pf (@pfs) {
+    $driver = script_output "lspci -v -s $pf | sed -n '/Kernel modules/p' | sed 's/.*Kernel modules: *//'";
+    my $is_known_driver = 0;
+    foreach (@drivers) {
+      next if ($_ ne "$driver");
+      $is_known_driver = 1;
+    }
+    push @drivers, $driver if ! $is_known_driver;
+  }
+  foreach (@drivers) {
+    assert_script_run("modprobe $driver", 60);
+  }
+ 
+  #set max_vfs and reload driver
+  #it seems the repeated enabling vf consused the system, or the some vf is not detached from a shutdown vm?
+  #so skip enabling in local test for debugging
+  record_info("Skip enabling VF");
+#  foreach $driver (@drivers) {
+#    assert_script_run("rmmod $driver", 30);   #4s last time
+#    assert_script_run("modprobe --first-time $driver max_vfs=8", 60);   #2s last time. but once fail with 30
+#  }
+
+  #bring up the SR-IOV device
+  foreach $pf (@pfs) {
+    my $nic = script_output "ls -l /sys/class/net |grep $pf | awk '{print \$9}'";
+    assert_script_run "echo \"BOOTPROTO='dhcp'\nSTARTMODE='manual'\" > /etc/sysconfig/network/ifcfg-$nic";
+    assert_script_run("ifup $nic", 60);   #15s last time
+  }
+  
+  my $vfs = script_output "lspci | grep Ethernet | grep \"Virtual Function\" | cut -d ' ' -f1";
+  my @vfs = split("\n", $vfs);
+
+}
+
+
+#set up guest test environment
+sub prepare_guest() {
+  my $vm = shift;
+
+  assert_script_run "virsh dumpxml $vm > $vm.xml";
+  script_run "virsh destroy $vm";
+  assert_script_run "virsh undefine $vm";
+
+  #extra process for XEN hypervisor
+  if (get_var("XEN") || check_var("SYSTEM_ROLE", "xen") || check_var("HOST_HYPERVISOR", "xen")) {
+
+    #enable pci-passthrough and set model for pv guest.
+    #refer to bug #1167217 for the reason
+    my $passthru_xml = "<passthrough state='on'/>";
+    my $e820_xml = "";
+    if ($vm =~ /\bpv\b/) {
+      $e820_xml = "<e820_host state='on'/>";
+    }
+    if (script_run("grep '<features>' $vm.xml") == 0) {
+      assert_script_run "sed -i \"/<features>/a\\<xen>\\n$passthru_xml\\n$e820_xml\\n</xen>\" $vm.xml";
+    }
+    else {
+      assert_script_run "sed -i \"/<domain /a\\<features>\\n<xen>\\n$passthru_xml\\n$e820_xml\\n</xen>\\n</features>\" $vm.xml";
```

Comment:
guest xml will have an empty line if $e820_xml is empty. So make sure it will cause any trouble

Reply:
Verified it is ok.

Reply:
Verified it is ok.